Add option to leave out definitions for common modules + namespace models

keithajackson opened the pull request:
When connecting to several LoopBack APIs from an Angular app, multiple definitions of LoopBackAuth and LoopBackResource will 'step on' each other, resulting in unpredictable behavior.

This change keeps the default behavior consistent, but will omit the module block that generates the LoopBackAuth, LoopBackAuthRequestInterceptor, and LoopBackResource components if a fourth parameter (excludeCommonModules) is set to 'true' when generateServices is called.

This will enable loopback-sdk-angular to more seamlessly be used with multiple-API architectures, without having to manually delete sections of code generated by the SDK.

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

keithajackson commented:
Rebased to match master. Please let me know if there are any desired changes and I'll be happy to accommodate.

@bajtos

bajtos commented:
@slnode ok to test

bajtos commented on lib/services.template.ejs:
```diff
@@ -374,7 +380,9 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
     }])
-
+<% } // end if (defineLoopBackAuthRequestInterceptor) %>
+defineLoopBackResource: <%= defineLoopBackResource %>
```
Hmm, what is the purpose of this line?

bajtos commented:
Hi @keithajackson, thank you for the pull request. Could you please add tests to verify your changes?

If I am reading the code correctly, when `defineLoopBackAuthRequestInterceptor` is true and `defineLoopBackResource` is false, the generated code will miss a semicolon at the end:

```js
module
    // ...
   .config(['$httpProvider', function($httpProvider) {
     $httpProvider.interceptors.push('LoopBackAuthRequestInterceptor');
   }])
   .factory('LoopBackAuthRequestInterceptor', ['$q', 'LoopBackAuth',
     function($q, LoopBackAuth) {
       // ...
     }])
//      ^-- missing semicolon
```

I am proposing to get rid of fine-grained flags `defineLoopBack*` and use a single flag instead.

bajtos commented on lib/services.js:
```diff
@@ -35,10 +44,23 @@ module.exports = function generateServices(app, ngModuleName, apiUrl) {
+  var defineLoopBackAuthRequestInterceptor = true;
+  var defineLoopBackResource = true;
+
+  if (excludeCommonModules === "true") {
```
Why a string? At this abstraction level, we should expect input arguments in the correct type and let callers to figure out how to convert string input into boolean values.

bajtos commented:
@keithajackson hello, what's the status of this pull request? Are you still keen to finish it?

keithajackson commented:
Hi @bajtos - my apologies for letting this go stale - I'm working on writing tests and catching up to master right now.

I'm running into some difficulty getting the test system up and running and was wondering if you might be able to give me a suggestion of what to try next.  I've been following the instructions in `DEVELOPING.md` to run tests with Karma.

When I run `node test.e2e/test-server` in a zsh terminal running Node 5.10.1, I get the output: `Test server is listening on http://localhost:3033`

The second terminal outputs:
```
[2016-06-17 14:41:37.563] [DEBUG] config - Loading config /Users/kjackson/Code/loopback-sdk-angular/karma.conf.js
17 06 2016 14:41:37.578:ERROR [runner]: There is no server listening on port 9876
```

I went into the `karma.conf.js` and changed the `port` from `9876` to `3033` to match the e2e server and re-ran `karma-start`.

The terminal running the test server outputs `POST /run 404 1.274 ms - 17`, but the terminal running `karma start` has another error:

```
[2016-06-17 14:42:38.113] [DEBUG] config - Loading config /Users/kjackson/Code/loopback-sdk-angular/karma.conf.js
Cannot POST /run
```

Does this issue sound familiar? Is there a server or Selenium webdriver I need to be running on port 9876 in addition to the test-server on 3033?

Thank you so much for your time and assistance!

bajtos commented:
Hey @keithajackson, I am sorry to hear about the difficulties with getting the tests environment set up. Before I dig deeper, I'd like to check whether `npm test` works for you?

keithajackson commented:
Hi @bajtos ,

I've added tests covering the new features, but  few builds are failing:

1. `default`: when I attempt to access the Jenkins server for details, I get an ERR_ADDRESS_UNREACHABLE error.

2. `loopback-sdk-angular.dependants`:

```
Set build name.
New build name is '#4244 (origin/master)'
Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#4244 (origin/master)', new value: '#4244 (origin/master)'
Notifying upstream projects of job completion
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Triggering a new build of update-commit-status
Finished: FAILURE
```

3. `pr-builder`:

```
Setting commit status on GitHub for https://github.com/strongloop/loopback-sdk-angular/commit/8d089fc41faf2a6f77abd9d67abc40de45319981
Notifying upstream projects of job completion
Warning: you have no plugins providing access control for builds, so falling back to legacy behavior of permitting any downstream builds to be triggered
Triggering a new build of update-commit-status
Finished: FAILURE
```

4. `dependent-loopback-workspace`:
```
  1) end-to-end empty-server template without explorer "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  2) end-to-end api-server template "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/es6-map/node_modules/es5-ext/test/number'
      at Error (native)

  3) end-to-end notes template "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  4) end-to-end hello-world template "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/math'
      at Error (native)

  5) end-to-end autoupdate "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  6) end-to-end discovery "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/test/array/#'
      at Error (native)

  7) end-to-end testConnection "before all" hook: installSandboxPackages:
     Error: timeout of 300000ms exceeded. Ensure the done() callback is being called in this test.
      at [object Object].<anonymous> (node_modules/loopback/node_modules/continuation-local-storage/node_modules/async-listener/glue.js:188:31)

  8) end-to-end start/stop/restart "before all" hook:
     Error: ENOTEMPTY, rmdir '/mnt/jenkins/workspace/loopback-workspace/1de32925/test/sandbox/node_modules/eslint/node_modules/escope/node_modules/es6-weak-map/node_modules/es5-ext/math/asinh'
      at Error (native)
```

Do you have any recommendations for how to address these issues?
Thanks!

bajtos commented on lib/services.js:
```diff
@@ -50,10 +50,14 @@ module.exports = function generateServices(app, options) {
       apiUrl: arguments[2],
     };
   }
+  var oldOptions = options;
```
AFAICT, this is unused - please remove.

keithajackson replied:
Ah yes - removed!

bajtos commented on lib/services.template.ejs:
```diff
@@ -37,8 +37,6 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
 <% for (var modelName in models) {
      var meta = models[modelName];
-     // capitalize the model name
-     modelName = modelName[0].toUpperCase() + modelName.slice(1);
```
`modelName` won't be available any more, are you sure it's not used anywhere in the template?

keithajackson replied:
`modelName` should be available inside the for loop, as this deleted code just updates the variable, rather than creating a new one (the formatting that used to be in the template is now in `getFormattedModelName` in `services.js`).  I double-checked that there isn't anywhere outside the for loop where modelName is used - it's just inside this for loop that we reference it directly.

bajtos commented on lib/services.template.ejs:
```diff
@@ -390,7 +389,6 @@ if (typeof module !== 'undefined' && typeof exports !== 'undefined' &&
         },
       };
     }])
-
```
Looks like unrelated whitespace change, please revert.

keithajackson replied:
Got it!

bajtos commented on test.e2e/given.js:
```diff
@@ -58,7 +58,9 @@ define(['angular', 'angularMocks', 'angularResource'], function(angular) {
    */
   given.servicesForLoopBackApp = function(options, cb) {
-    options.name = generateUniqueServiceName(options.name);
+    if (! options.name) {
```
style:

```js
if (!options.name) {
```

keithajackson replied:
Sounds good!

bajtos commented on test.e2e/test-server.js:
```diff
@@ -105,19 +104,8 @@ masterApp.post('/setup', function(req, res, next) {
-        // the old API, test it to verify backwards compatibility
-        servicesScript = generator.services(lbApp, name, apiUrl);
-      }
+      servicesScript = generateService(generator, lbApp, apiUrl, opts);
```
Please keep "the old API" branch, we need to keep it covered by unit-tests to prevent regressions.

bajtos replied:
See https://github.com/strongloop/loopback-sdk-angular/pull/224#discussion_r72442921

bajtos replied:
Oh, I see, this is handled by `generateService`, ignore my comment then.

bajtos commented on test.e2e/test-server.js:
```diff
@@ -161,6 +149,27 @@ masterApp.listen(port, function() {
 
+function generateService(generator, lbApp, apiUrl, opts) {
+  var servicesScript;
+  if (typeof opts.includeSchema !== 'undefined' ||
```
In Loopback, we use simpler `opts.includeSchema !== undefined`. Please simplify here and on the lines below too.

keithajackson replied:
Done!

bajtos commented on test.e2e/test-server.js:
```diff
@@ -161,6 +149,27 @@ masterApp.listen(port, function() {
+
+    // build options object for new options-based API
+    var generatorOptions = opts;
+    generatorOptions.ngModuleName = opts.name;
```
I think you are missing `apiUrl` here?

keithajackson replied:
Ah yes! Added - thanks!

bajtos commented:
Hey @keithajackson, sorry for the long delay. I reviewed your changes, they looks mostly good to me. Please address the comments I left above.

As for failing CI builds, I am afraid our CI system is still not stable and still not visible to public. Just make sure `npm test` passes on your local machine, we will check CI results ourselves to see if they are related to this patch.

slnode commented:
Can one of the admins verify this patch?

keithajackson commented:
@bajtos All comments adressed and code rebased against master - thanks so much for close review! Let me know if you have any other comments you'd like addressed or run into any surprises.

I double-checked test after the update+rebase and it all looks good on my machine.

bajtos commented:
Thank you for the quick update, @keithajackson!

I run `node apidocs/describe-builtin-models.js` using `master` version and the version from your patch. When I compared the outcome, there is one extra empty line:

```diff
--- apidocs/loopback-core-master.js	2016-07-28 14:43:03.000000000 +0200
+++ apidocs/loopback-core.js	2016-07-28 14:43:18.000000000 +0200
@@ -6993,6 +6993,7 @@
       }]);


+
   module
   .factory('LoopBackAuth', function() {
     var props = ['accessTokenId', 'currentUserId', 'rememberMe'];
```

What's your opinion? Do you feel like fixing this small detail, or should we ignore it?

Other than that, the patch LGTM.

Could you please squash all your commits into a single one and provide a descriptive commit message following the [50/72 rule](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)?

keithajackson commented:
@bajtos Fixed! I was able to find the blank line causing the documentation inconsistency and removed it.  I've also squashed commits to match the commit style.

Thanks again for your time and thoughtful comments. Let me know if there is anything else I can improve or help with!

bajtos commented:
@keithajackson awesome! I have landed the patch, thank you for this great contribution 🎉 👏

bajtos commented:
Released as `loopback-sdk-angular@1.9.0`, enjoy 😄
